lib/segment/src/common/mmap_ops.rs

```diff
@@ -0,0 +1,137 @@
+use std::fs::OpenOptions;
+use std::mem::size_of;
+use std::ops::DerefMut;
+use std::path::Path;
+use std::{mem, slice};
+
+use memmap2::{Mmap, MmapMut, MmapOptions};
+
+use crate::entry::entry_point::OperationResult;
+use crate::madvise;
+
+pub fn create_and_ensure_length(path: &Path, length: usize) -> OperationResult<()> {
+    let file = OpenOptions::new()
+        .read(true)
+        .write(true)
+        .create(true)
+        .open(path)?;
+
+    file.set_len(length as u64)?;
+    Ok(())
+}
+
+pub fn open_read_mmap(path: &Path) -> OperationResult<Mmap> {
+    let file = OpenOptions::new()
+        .read(true)
+        .write(false)
+        .append(true)
+        .create(true)
+        .open(path)?;
+
+    let mmap = unsafe { MmapOptions::new().map(&file)? };
+    madvise::madvise(&mmap, madvise::get_global())?;
+    Ok(mmap)
+}
+
+pub fn open_write_mmap(path: &Path) -> OperationResult<MmapMut> {
+    let file = OpenOptions::new()
+        .read(true)
+        .write(true)
+        .create(false)
+        .open(path)?;
+
+    let mmap = unsafe { MmapMut::map_mut(&file)? };
+    madvise::madvise(&mmap, madvise::get_global())?;
+    Ok(mmap)
+}
+
+pub fn transmute_from_u8<T>(data: &[u8]) -> &T {
+    debug_assert!(data.len() == size_of::<T>());
+    let ptr = data.as_ptr() as *const T;
+    unsafe { &*ptr }
+}
+
+pub fn transmute_from_u8_to_mut<'a, T>(data: &[u8]) -> &'a mut T {
```

Comment:
This is UB. See the rust nomicon: https://doc.rust-lang.org/nomicon/transmutes.html


> Transmuting an & to &mut is Undefined Behavior. While certain usages may appear safe, note that the Rust optimizer is free to assume that a shared reference won't change through its lifetime and thus such transmutation will run afoul of those assumptions. So:
>
>Transmuting an & to &mut is always Undefined Behavior.
No you can't do it.
No you're not special.

Reply:
i think we can do mut to mut herr